api/opensource/mfx_dispatch/linux/mfxloader.cpp

```diff
@@ -0,0 +1,582 @@
+        }
+
+        if (m_version < par.Version) {
+          mfx_res = MFX_ERR_UNSUPPORTED;
+          break;
+        }
+
+        mfx_res = ((decltype(MFXQueryIMPL)*)m_table[eMFXQueryIMPL])(m_session, &m_implementation);
+        if (MFX_ERR_NONE != mfx_res) {
+          mfx_res = MFX_ERR_UNSUPPORTED;
+          break;
+        }
+      } while(false);
+
+      if (MFX_ERR_NONE == mfx_res) {
+        m_dlh = hdl;
+        break;
+      } else {
+        Close();
+      }
+    }
+  }
+
+  return mfx_res;
+}
+
+mfxStatus LoaderCtx::Close()
+{
+  auto proc = (decltype(MFXClose)*)m_table[eMFXClose];
+  mfxStatus mfx_res = (proc)? (*proc)(m_session): MFX_ERR_NONE;
+
+  m_implementation = {};
+  m_version = {};
+  m_session = nullptr;
+  memset(&m_table, 0, sizeof(m_table));
+  return mfx_res;
+}
+
+mfxStatus PluginCtx::Load(const mfxPluginUID& uid, mfxU32 version, const char *path)
+{
+  if (!path) {
+    return MFX_ERR_NULL_PTR;
+  }
+
+  mfxStatus mfx_res = MFX_ERR_NONE;
+  std::shared_ptr<void> hdl = make_dlopen(path, RTLD_LOCAL|RTLD_NOW);
+
+  if (!hdl.get()) {
+    return MFX_ERR_NOT_FOUND;
+  }
+
+  do {
+    m_uid = uid;
+    m_create_plugin = (CreatePluginPtr)dlsym(hdl.get(), "CreatePlugin");
+    if (!m_create_plugin) {
+      mfx_res = MFX_ERR_NOT_FOUND;
+      break;
+    }
+
+    mfx_res = m_create_plugin(m_uid, &m_plugin);
+    if (MFX_ERR_NONE != mfx_res) {
+      break;
+    }
+
+    mfx_res = m_plugin.GetPluginParam(m_plugin.pthis, &m_plugin_param);
+    if (MFX_ERR_NONE != mfx_res) {
+      break;
+    }
+
+    mfx_res = MFXVideoUSER_Register((mfxSession)&m_loader, m_plugin_param.Type, &m_plugin);
+    if (MFX_ERR_NONE != mfx_res) {
+      break;
+    }
+  } while(false);
+
+  if (MFX_ERR_NONE == mfx_res) {
+    m_dlh = hdl;
+  } else {
+    m_uid = {};
+    m_create_plugin = nullptr;
+    m_plugin = {};
+    m_plugin_param = {};
+  }
+  return mfx_res;
+}
+
+mfxStatus PluginCtx::Unload()
+{
+  return MFXVideoUSER_Unregister((mfxSession)&m_loader, m_plugin_param.Type);
+}
+
+mfxStatus LoaderCtx::LoadPlugin(const mfxPluginUID &uid, mfxU32 version, const char *path)
+{
+  if (!path) return MFX_ERR_NULL_PTR;
+
+  try {
+    std::lock_guard<std::mutex> lock(m_guard);
+    for (auto it: m_plugins) {
+      if (it.getUID() == uid) return MFX_ERR_UNDEFINED_BEHAVIOR;
+    }
+  } catch(...) {
+    return MFX_ERR_UNKNOWN;
+  }
+
+  mfxStatus mfx_res = MFX_ERR_NONE;
+  std::list<PluginCtx> ctx;
+
+  try {
+    ctx.emplace_front(PluginCtx(*this));
+  } catch(...) {
+    return MFX_ERR_UNKNOWN;
+  }
+
+  mfx_res = ctx.front().Load(uid, version, path);
+  if (MFX_ERR_NONE != mfx_res) {
+    return mfx_res;
+  }
+
+  try {
+    std::lock_guard<std::mutex> lock(m_guard);
+    m_plugins.splice(m_plugins.end(), ctx);
+  } catch(...) {
+    return MFX_ERR_UNKNOWN;
+  }
+
+  return MFX_ERR_NONE;
+}
+
+mfxStatus LoaderCtx::UnloadPlugin(const mfxPluginUID& uid)
+{
+  mfxStatus mfx_res = MFX_ERR_NONE;
+
+  try {
+    std::list<PluginCtx> ctx;
```

Comment:
It is complicated on purpose, I think you did not catch the idea which is to move bottom half of plugin ctx destroy (which may contain potentially long dlclose) out of the mutex protection. I will add the comment and will use std::find_if instead of the loop, but splice() will remain.

Reply:
ok, got it